### PR TITLE
USHIFT-156: csi plugin implementation

### PR DESCRIPTION
**Pull request by copejon**
This PR refactors the default storage provisioner by removing the hostpath provisioner and replacing it with the lvm csi plugin.

**Which issue(s) this PR addresses**:
[USHIFT-156](https://issues.redhat.com//browse/USHIFT-156)

### CODE REVIEW

File: assets/components/odf-lvm/topolvm-openshift-storage_namespace.yaml
Comment (dhellmann) [inline assets/components/odf-lvm/topolvm-openshift-storage_namespace.yaml:10]: I don't understand what this is doing. Why do we need so many things to have this label and why shouldn't openshift or kube-infra namespaces be seen by TopoLVM?

File: assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml
Comment (dhellmann) [inline assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml:4]: What's the current schedule for this API changing? We should make sure we have a jira ticket to figure out how to handle the transition within MicroShift.

File: pkg/components/storage.go
Comment (dhellmann) [inline pkg/components/storage.go:61]: Were these panic changes left over from debugging?

File: assets/components/odf-lvm/topolvm-controller_deployment.yaml
Comment (dhellmann) [reply assets/components/odf-lvm/topolvm-controller_deployment.yaml:86]: Using a string template can wait for another PR

File: assets/components/odf-lvm/topolvm-controller_deployment.yaml
Comment (copejon) [reply assets/components/odf-lvm/topolvm-controller_deployment.yaml:70]: I believe this is related to [CSIStorageCapacity](https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/csi-storage-capacity-v1/) tracking.  This feature allows us to exclude the topolvm extended scheduler.

File: pkg/components/storage.go
Comment (copejon) [reply pkg/components/storage.go]: fixed

**Comment by copejon** [comment]
Holding to allow for reviews

/hold

File: assets/components/odf-lvm/topolvm-controller_deployment.yaml
Comment (copejon) [reply assets/components/odf-lvm/topolvm-controller_deployment.yaml]: Fixed 👍🏻

File: assets/components/odf-lvm/topolvm-openshift-storage_namespace.yaml
Comment (copejon) [reply assets/components/odf-lvm/topolvm-openshift-storage_namespace.yaml:10]: Actually, I believe this is only a requirement when running the extended scheduler.  Will remove.

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml
Comment (dhellmann) [inline assets/components/odf-lvm/topolvm_default-storage-class.yaml:9]: Is this something users will eventually want to configure here or would we suggest that they create a separate StorageClass to do that?

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml
Comment (dhellmann) [inline assets/components/odf-lvm/topolvm_default-storage-class.yaml:7]: Is this name significant to the code? Should it be a variation of ODF-LVM to be consistent with branding elsewhere?

File: assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml
Comment (copejon) [reply assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml:4]: The API's groupVersionKind is expected to change in or by Q4 this year to replace the maintaining company's name.  There's not a set date for this yet.  General versioning has a spike already, but it'll be worth it to have this upcoming change tracked for awareness I think.

File: assets/components/odf-lvm/topolvm-controller_deployment.yaml
Comment (dhellmann) [reply assets/components/odf-lvm/topolvm-controller_deployment.yaml:70]: Does it need to run so often? Is it something users are likely going to need to be able to configure to manage the performance of the system?

File: assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml
Comment (dhellmann) [reply assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml:4]: I'll be very interested to hear what the ODF team's plan is for rolling upgrades with a GVK name change like that.

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml
Comment (copejon) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:9]: The storageClass is typically a cluster-admin defined API, so it's reasonable to assume to microshift users would want to specify this.

Comment (dhellmann) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:9]: OK, so we don't need to provide a config option for this, but we might need to give documentation for how to create an alternative storage class.

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml
Comment (copejon) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:7]: No, the storageClass name can be arbitrary, as can the deployment and daemonset, and other included APIs.  

WRT branding, on OCP the ODF LVM Operator deploys all the operand APIs (pods, configMaps, secrets, role, roleBindings, etc) with a `topolvm-` prefix.  Functionally, this is a trivial string replacement in rebase automation.

I'm a little concerned about renaming a OCP component for the sake of branding, which could create the appearance of different LVM provisioning solutions between OCP and MicroShift.  I'll defer to your take on that though.

Comment (dhellmann) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:7]: If the LVMO uses the `topolvm-` prefix elsewhere, I agree we should stay consistent with that. Thanks for explaining the reasoning here. We should try to mention this decision in the enhancement as well.

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml
Comment (copejon) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:9]: +1

File: assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml
Comment (copejon) [reply assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml:4]: I'm following up with them internally.  Upstream, there's a plan to implement a migration tool, but I don't know what plans for downstream users will be.

File: assets/components/odf-lvm/topolvm-node_daemonset.yaml
Comment (mangelajo) [inline assets/components/odf-lvm/topolvm-node_daemonset.yaml:23]: image reference templates here too.

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml
Comment (mangelajo) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:7]: Maintaining consistency makes sense to me too

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml
Comment (mangelajo) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:9]: :+1

File: assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml
Comment (mangelajo) [reply assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml:4]: May be a web hook in the meanwhile to provide translation between both versions?

File: pkg/release/release_arm64.go
Comment (mangelajo) [inline pkg/release/release_arm64.go:29]: If you ask @fzdarsky he can share how to find the arm64 images.

**Review by mangelajo — CHANGES_REQUESTED** [review]
some tiny comments, other than that it looks good

File: assets/components/odf-lvm/topolvm-controller_deployment.yaml
Comment (dhellmann) [reply assets/components/odf-lvm/topolvm-controller_deployment.yaml:83]: I'd be OK doing that in a follow-up PR if that helps us keep moving.

**Comment by dhellmann** [comment]
/lgtm

Let's take this and deal with the cleanup items as follow-ups.

**Pull request merged**
